go/engine/team_post.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package engine
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+
+	"golang.org/x/crypto/nacl/box"
+)
+
+type NewTeamEngine struct {
+	libkb.Contextified
+	name string
+}
+
+func NewNewTeamEngine(g *libkb.GlobalContext, name string) *NewTeamEngine {
+	return &NewTeamEngine{
+		Contextified: libkb.NewContextified(g),
+		name:         name,
+	}
+}
+
+func (e *NewTeamEngine) Name() string {
+	return "NewTeam"
+}
+
+func (e *NewTeamEngine) Prereqs() Prereqs {
+	return Prereqs{
+		Device: true,
+	}
+}
+
+func (e *NewTeamEngine) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{
+		libkb.LogUIKind,
+		libkb.SecretUIKind,
+	}
+}
+
+func (e *NewTeamEngine) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{}
+}
+
+func (e *NewTeamEngine) Run(ctx *Context) (err error) {
+	defer e.G().Trace("NewTeamEngine", func() error { return err })()
+
+	me, err := libkb.LoadMe(libkb.NewLoadUserArg(e.G()))
+	if err != nil {
+		return err
+	}
+
+	ska := libkb.SecretKeyArg{
+		Me:      me,
+		KeyType: libkb.DeviceSigningKeyType,
+	}
+	sigKey, err := e.G().Keyrings.GetSecretKeyWithPrompt(ctx.SecretKeyPromptArg(ska, "to create a new team"))
+	if err != nil {
+		return err
+	}
+	if err = sigKey.CheckSecretKey(); err != nil {
+		return err
+	}
+
+	ownerName, err := libkb.NameWithEldestSeqno(me.GetName(), me.GetCurrentEldestSeqno())
+	// An error happens here if the seqno isn't loaded for some reason.
+	if err != nil {
+		return err
+	}
+
+	id := libkb.RootTeamIDFromName(e.name)
+	teamSection := libkb.TeamSection{
+		Name: e.name,
+		ID:   id,
+	}
+	teamSection.Members.Owner = []string{ownerName}
+	teamSection.Members.Admin = []string{}
+	teamSection.Members.Writer = []string{}
+	teamSection.Members.Reader = []string{}
+	teamSection.SharedKey.Boxes = map[string]string{}
+
+	ephemeralPair, err := libkb.GenerateNaclDHKeyPair()
+	if err != nil {
+		return err
+	}
+	teamSection.SharedKey.E = ephemeralPair.Public.GetKID().String()
+	teamSection.SharedKey.Gen = 1
+
+	// This is the shared secret key. Do not store it except encrypted in the
+	// boxes that follow!
+	sharedSecretKey, err := libkb.RandBytes(32)
+	if err != nil {
+		return err
+	}
+
+	// Note that the key boxes use usernames directly, without the %Seqno
+	// annotation from the roles section.
```

Comment:
@maxtaco Will address the rest of the refactoring in a follow-up commit, but curious to get your take on this note in particular. The server returns an error if I try to use `username%seqno` as a key in the boxes list, because what ends up in `_put_team_member_box` hasn't been parsed that way. Expected?